applications/AdjointFluidApplication/custom_utilities/numerical_diffusion.h

```diff
@@ -0,0 +1,438 @@
+
+        KRATOS_CATCH("")
+    }
+
+    template<unsigned int TDim>
+    void CalculateVelocityGradientTensor(
+        Element::Pointer pElement,
+        const ProcessInfo& rCurrentProcessInfo,
+        MatrixType& rMatrix
+    )
+    {
+        constexpr unsigned int TNumNodes = TDim + 1;        
+
+        if (rMatrix.size1() != TDim || rMatrix.size2() != TDim)
+            rMatrix.resize(TDim,TDim,false);
+        
+        rMatrix.clear();
+
+        boost::numeric::ublas::bounded_matrix<double, TNumNodes, TDim> DN_DX;
+        array_1d< double, TNumNodes > N;
+        double Volume;
+        GeometryUtils::CalculateGeometryData(pElement->GetGeometry(),DN_DX,N,Volume);
+
+        std::vector<array_1d<double, TDim>> rNodalVelocityVectors;
+        for (unsigned int iNode = 0; iNode < TNumNodes; iNode++)
+            rNodalVelocityVectors.push_back(pElement->GetGeometry()[iNode].FastGetSolutionStepValue(VELOCITY));
+
+        for (unsigned int i = 0; i < TDim; i++)
+            for (unsigned int j = 0; j < TDim; j++)
+                for (unsigned int k = 0; k < TNumNodes; k++)
+                    rMatrix(i,j) += DN_DX(k,i)*rNodalVelocityVectors[k][j];
+
+    }
+
+    template<unsigned int TDim>
+    double CalculateNumericalDiffusionSVMethodPressureCoupled(Element::Pointer pCurrentElement, const ProcessInfo& rCurrentProcessInfo)
+    {
+        constexpr unsigned int TNumNodes = TDim + 1;        
+
+        boost::numeric::ublas::bounded_matrix<double, TNumNodes, TDim> DN_DX;
+        array_1d< double, TNumNodes > N;
+        double Volume;
+
+        GeometryType& rGeom = pCurrentElement->GetGeometry();
+
+        GeometryUtils::CalculateGeometryData(rGeom,DN_DX,N,Volume);
+
+        MatrixType GradVel;
+
+        CalculateVelocityGradientTensor<TDim>(pCurrentElement, rCurrentProcessInfo, GradVel);
+
+        double velocity_divergence = 0.0;
+        for (IndexType i=0;i<TDim; i++)
+            velocity_divergence += GradVel(i,i);
+        
+        Eigen::Matrix<double, TDim+1, TDim+1>  M;
+        for (IndexType i=0; i < TDim; i++)
+            M(i,i) = 0.5 * velocity_divergence - GradVel(i,i);
+        for (IndexType i=0; i < TDim; i++)
+            for (IndexType j=i+1; j < TDim; j++)
+            {
+                M(i,j) = -GradVel(i,j);
+                M(j,i) = -GradVel(j,i);
+            }
+        for (IndexType i=0; i < TDim + 1; i++)
+        {
+            M(TDim, i) = 0.0;
+            M(i, TDim) = 0.0;
+        }
+        M(TDim, TDim) = 0.5 * velocity_divergence;
+
+        Eigen::JacobiSVD<Eigen::Matrix<double, TDim+1, TDim+1>> svd(M, Eigen::ComputeThinU | Eigen::ComputeThinV);
+
+        const auto& S = svd.singularValues();
+
+        double numerical_viscosity;
+        numerical_viscosity = S[0];
+
+        return numerical_viscosity;
+    }
+
+    template<unsigned int TDim>
+    double CalculateNumericalDiffusionSVMethodPressureDecoupled(Element::Pointer pCurrentElement, const ProcessInfo& rCurrentProcessInfo)
```

Comment:
This function is nearly identical to the PressureCoupled. These should be refactored to remove code duplication and make it more clear to the reader where the actual difference is.